### PR TITLE
cleanup mitmproxy.controller, raise Kill in Channel

### LINKED ISSUE #907: On Shutdown: flow is None and crashes
The Version is 0.15. Crash info as follow:

```
./mitmdump.run.sh: line 2: 14504 Segmentation fault      mitmdump -p 59873 > network
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/libmproxy/proxy/server.py", line 120, in handle
    root_layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/proxy/modes/http_proxy.py", line 10, in __call__
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/tls.py", line 272, in __call__
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 130, in __call__
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 323, in __call__
    self.handle_regular_mode_connect(request)
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 409, in handle_regular_mode_connect
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/tls.py", line 272, in __call__
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 130, in __call__
    layer()
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 340, in __call__
    self.get_response_from_server(flow)
  File "/usr/local/lib/python2.7/site-packages/libmproxy/protocol/http.py", line 471, in get_response_from_server
    if flow.response.stream:
AttributeError: 'NoneType' object has no attribute 'response'
```

### PR DESCRIPTION
**Pull request by mhils**
This PR refactors `controller.Master` into `controller.Master` and `controller.ServerMaster` and cleans up a couple of inconsistencies. Also, `Channel` now raises a `Kill` exception when the server shuts down, which fixes #907.

@Kriechi: How understandable is `mitmproxy.controller` now? My primary focus was to make this more readable, did I succeed at least a bit? (You'd need to look at the file probably, the diff is fairly broken)

### CODE REVIEW

**Comment by Kriechi** [comment]
LGTM
I guess this superseeds somehow #1057?

**Comment by mhils** [comment]
I think so, yes. #1057 is really hard to follow for me - we may want to keep at least the tests from there, #1084 is also somehow related. Whatsoever, I'm going to land this now.
